spec/acceptance/throttling_spec.rb

```diff
@@ -34,6 +34,27 @@
     end
   end
 
+  it "returns correct Retry-After header if enabled" do
+    begin
+      prev = Rack::Attack.retry_after_header_enabled
+      Rack::Attack.retry_after_header_enabled = true
+
+      Rack::Attack.throttle("by ip", limit: 1, period: 60) do |request|
+        request.ip
+      end
+
+      get "/", {}, "REMOTE_ADDR" => "1.2.3.4"
+      assert_equal 200, last_response.status
+
+      Timecop.travel(25) do
+        get "/", {}, "REMOTE_ADDR" => "1.2.3.4"
+        assert last_response.headers["Retry-After"].to_i <= 60 - 25
```

Comment:
As far as I understand rack-attack internals, this can not not be done exactly due to how expiration time is calculated https://github.com/kickstarter/rack-attack/blob/9cf227a32ec1b3e122166f00ee72e6570f305c5d/lib/rack/attack/cache.rb#L46-L51
as this depends on which exact second/minute this is calculated.
Or am I missing something and then how this can be done?

Reply:
Yeah, tricky.

But I think it can be done by making the test run aligned with epoch = 0.

```diff
diff --git a/spec/acceptance/throttling_spec.rb b/spec/acceptance/throttling_spec.rb
index a9021c6..20b5882 100644
--- a/spec/acceptance/throttling_spec.rb
+++ b/spec/acceptance/throttling_spec.rb
@@ -43,13 +43,17 @@ describe "#throttle" do
         request.ip
       end
 
-      get "/", {}, "REMOTE_ADDR" => "1.2.3.4"
+      Timecop.freeze(Time.at(0)) do
+        get "/", {}, "REMOTE_ADDR" => "1.2.3.4"
+      end
+
       assert_equal 200, last_response.status
 
-      Timecop.travel(25) do
+      Timecop.freeze(Time.at(25)) do
         get "/", {}, "REMOTE_ADDR" => "1.2.3.4"
-        assert last_response.headers["Retry-After"].to_i <= 60 - 25
       end
+
+      assert_equal "35", last_response.headers["Retry-After"]
     ensure
       Rack::Attack.throttled_response_retry_after_header = prev
     end
```

Reply:
Cleverly 👍